packages/hash/datastore/postgres/migration/1649694034320_aggregations-add-source-timestamps.ts

```diff
@@ -0,0 +1,140 @@
+    set
+      applied_to_source_at = (
+        select entity_versions.updated_at
+        from entity_versions
+        where entity_versions.entity_version_id = ANY(aggregations.source_entity_version_ids)
+        order by entity_versions.updated_at
+        asc
+        limit 1
+      )
+    where applied_to_source_at is null
+  `);
+  pgm.sql(`
+    update aggregations
+    set
+      applied_to_source_by_account_id = (
+        select entity_versions.updated_by_account_id
+        from entity_versions
+        where entity_versions.entity_version_id = ANY(aggregations.source_entity_version_ids)
+        order by entity_versions.updated_at
+        asc
+        limit 1
+      )
+    where applied_to_source_by_account_id is null
+  `);
+
+  /**
+   * Step 3. populate removed_from_source_at/by_account_id for existing aggregations
+   */
+  pgm.sql(`
+   update aggregations
+   set
+    removed_from_source_at = (
+      select entity_versions.updated_at
+      from entity_versions
+      where entity_versions.entity_version_id = ANY(aggregations.source_entity_version_ids)
+      order by entity_versions.updated_at
+      desc
+      limit 1
+    ),
+    removed_from_source_by_account_id = (
+      select entity_versions.updated_by_account_id
+      from entity_versions
+      where entity_versions.entity_version_id = ANY(aggregations.source_entity_version_ids)
+      order by entity_versions.updated_at
+      desc
+      limit 1
+    )
+    where
+        removed_from_source_at is null
+      and (
+        select entity_versions.entity_version_id
+        from entity_versions
+        where entity_versions.entity_version_id = ANY(aggregations.source_entity_version_ids)
+        order by entity_versions.updated_at
+        desc
+        limit 1
+      ) != (
+        select entity_versions.entity_version_id
+        from entity_versions, aggregations
+        where
+          entity_id = (
+            select entity_versions.entity_id
+            from entity_versions
+            where entity_versions.entity_version_id = ANY(aggregations.source_entity_version_ids)
+            order by entity_versions.updated_at
+            desc
+            limit 1
+          )
+        order by entity_versions.updated_at
+        desc
+        limit 1
+      )
+ `);
+
+  /**
+   * Step 4. make applied_to_source_at/by_account_id fields non-nullable
+   */
+  pgm.alterColumn("aggregations", "applied_to_source_at", { notNull: true });
+  pgm.alterColumn("aggregations", "applied_to_source_by_account_id", {
+    notNull: true,
+  });
+
+  /**
+   * Step 5. drop the source_entity_version_ids column
+   */
+  pgm.dropColumns(
+    "aggregations",
+    ["source_entity_version_ids", "created_at", "created_by_account_id"],
+    { ifExists: true },
+  );
+}
+
+/**
+ * Implementing a down migration that wouldn't incur dataloss is possible, but
+ * would involve significant effort deemed out of scope at this stage, so has been left unimplemented.
```

Comment:
Just an edit to clarify that we will never do this

```suggestion
 * would involve significant effort deemed out of scope at this stage of the project.
 * It will be left unimplemented.
```

Reply:
If a user came along and wanted to implement this would we complain, or would it have any utility at all? I'm wondering if we can start tagging "unimplemented and low priority, but if you want to do this knock yourself out" tickets inline in code somehow (ideally then easily discoverable through some unique identifier along the lines of our `@todo` tags).

Reply:
I would tell them no, because it would be a complete waste of their time and ours getting it merged.

For context, the purpose of the down migration is to undo the change to the db, because you've deployed it and something has gone wrong, or because you want to migrate it back and forth locally for some reason.

Once the change is merged in, deployed, and nothing has gone wrong, no-one is ever going to run a down migration somewhere back in the history. Its utility is fleeting.

Reply:
I guess in theory you could have some future world in which we wanted to surgically undo some db change, and there happened to be a down migration in the history that did that, on the assumption that all the db changes since had not made that a complete mess. But it seems unlikely.

Reply:
I don't think this will ever serve a purpose if it's implemented, so think it makes most sense to not encourage others to do so with stronger wording that it "will be left unimplemented".